lib/logproto/logproto-proxied-text-server.c

```diff
@@ -0,0 +1,290 @@
+ * COPYING for details.
+ *
+ */
+
+
+#include <stdio.h>
+#include <stdlib.h>
+#include "messages.h"
+#include "logproto-proxied-text-server.h"
+#include "str-utils.h"
+
+#define PROXY_HDR_TCP4 "PROXY TCP4 "
+#define PROXY_HDR_TCP6 "PROXY TCP6 "
+#define PROXY_HDR_UNKNOWN "PROXY UNKNOWN"
+#define PROXY_PROTO_HDR_MAX_LEN_RFC 108
+#define PROXY_PROTO_HDR_MAX_LEN (PROXY_PROTO_HDR_MAX_LEN_RFC * 2)
+
+static gboolean
+_check_header_length(const guchar *msg, gsize msg_len)
+{
+  if (msg_len > PROXY_PROTO_HDR_MAX_LEN_RFC)
+    {
+      msg_warning("PROXY proto header length exceeds max. length defined by specification",
+                  evt_tag_long("length", msg_len),
+                  evt_tag_str("header", (const gchar *)msg));
+    }
+
+  if (msg_len > PROXY_PROTO_HDR_MAX_LEN)
+    {
+      msg_error("PROXY proto header with invalid header length",
+                evt_tag_int("max_parsable_length", PROXY_PROTO_HDR_MAX_LEN),
+                evt_tag_int("max_length_by_spec", PROXY_PROTO_HDR_MAX_LEN_RFC),
+                evt_tag_long("length", msg_len),
+                evt_tag_str("header", (const gchar *)msg));
+      return FALSE;
+    }
+
+  return TRUE;
+}
+
+static gboolean
+_check_header(const guchar *msg, gsize msg_len, const gchar *expected_header, gsize *header_len)
+{
+  gsize expected_header_length = strlen(expected_header);
+
+  if (msg_len < expected_header_length)
+    return FALSE;
+
+  *header_len = expected_header_length;
+  return strncmp((const gchar *)msg, expected_header, expected_header_length) == 0;
+}
+
+static gboolean
+_is_proxy_proto_tcp4(const guchar *msg, gsize msg_len, gsize *header_len)
+{
+  return _check_header(msg, msg_len, PROXY_HDR_TCP4, header_len);
+}
+
+static gboolean
+_is_proxy_proto_tcp6(const guchar *msg, gsize msg_len, gsize *header_len)
+{
+  return _check_header(msg, msg_len, PROXY_HDR_TCP6, header_len);
+}
+
+static gboolean
+_is_proxy_unknown(const guchar *msg, gsize msg_len, gsize *header_len)
+{
+  return _check_header(msg, msg_len, PROXY_HDR_UNKNOWN, header_len);
+}
+
+static gboolean
+_parse_unknown_header(LogProtoProxiedTextServer *self, const guchar *msg, gsize msg_len)
+{
+  if (msg_len == 0)
+    return TRUE;
+
+  GString *params_str = g_string_new_len((const gchar *)msg, msg_len);
+  msg_warning("PROXY UNKNOWN header contains unexpected paramters", evt_tag_str("parameters", params_str->str));
+  return TRUE;
+}
+
+static gboolean
+_parse_tcp_header(LogProtoProxiedTextServer *self, const guchar *msg, gsize msg_len)
+{
+  if (msg_len == 0)
+    return FALSE;
+
+  GString *params_str = g_string_new_len((const gchar *)msg, msg_len);
+  gboolean result = FALSE;
+  msg_debug("PROXY header params", evt_tag_str("params", (const gchar *)msg));
+
+  gchar **params = strsplit(params_str->str, ' ', 5);
+  gint params_n = g_strv_length(params);
+  if (params_n < 4)
+    goto ret;
+
+  strncpy(self->info->src_ip, params[0], IP_BUF_SIZE - 1);
```

Comment:
`strncpy` will not null-terminate `src_ip` in case `params[0]` is longer than `IP_BUF_SIZE - 1`, but the `info` struct is created with `g_new0`, so there will be an implicit \0 at the end of each buffer, which is perfect. 👍🏻